### PR TITLE
Added DAL tests project

### PR DESCRIPTION
**Pull request by brtkkndrl**
Created DAL project.
Updated Repository naming;
Updated IUnitOfWork;
Added for updating entity, and removed SaveChanges from Repository.

### CODE REVIEW

**Comment by vecerap** [comment]
@BRTEK007 Why this is a draft?

**Comment by brtkkndrl** [comment]
> @BRTEK007 Why this is a draft?
I thought here would be more discussion, marked it for review.

**Review by KoubaFromPucks — APPROVED** [review]

File: src/Quizzer.API.DAL/Repository/Repository.cs
Comment (vecerap) [inline src/Quizzer.API.DAL/Repository/Repository.cs:1]: Why did you change the name?

File: src/Quizzer.API.DAL.Tests/ApiDALTests.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+            Guid? repoEntityId = repo.Update(updatedEntity);
+            await UnitOfWork.CommitAsync();
+
+            //Assert
+            Assert.NotNull(repoEntityId);
+            Assert.Equal(repoEntityId, updatedEntity.Id);
+            await using QuizzerDbContext dbx = await DbContextFactory.CreateDbContextAsync();
+            QuizEntity dbxEntity = await dbx.Quiezzes.SingleAsync(i => i.Id == repoEntityId);
+            DeepAssert.Equal<QuizEntity>(updatedEntity, dbxEntity);
+        }
```
Comment (vecerap) [inline src/Quizzer.API.DAL.Tests/ApiDALTests.cs]: I assume this is the test, which validates the saves changes inside the repository which was removed?

File: src/Quizzer.DAL.Tests/RepositoryTests.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+            UnitOfWork = UnitOfWorkFactory.Create();
+        }
+
+        [Fact]
+        public void Test1() {
+
+            var repo = UnitOfWork.GetRepository<QuizEntity>();
+
+            //Assert.
+        }
```
Comment (vecerap) [inline src/Quizzer.DAL.Tests/RepositoryTests.cs]: This should not be here, also what is then the purpose of this file?

File: src/Quizzer.API.DAL/UnitOfWork/IUnitOfWork.cs [hunk 3]
```diff
@@ -11,10 +13,25 @@ public interface IUnitOfWork : IDisposable {
...
+
+        /// <summary>
+        /// Makes changes on entities in database
+        /// </summary>
+        public Task CommitAsync();
+
+        /// <summary>
+        /// Discards changes
+        /// </summary>
+        public ValueTask DisposeAsync();
```
Comment (vecerap) [inline src/Quizzer.API.DAL/UnitOfWork/IUnitOfWork.cs]: I do not see you are using `IAsyncDisposable` interface in uow, the runtime will probably not trigger this method...

**Review by vecerap — COMMENTED** [review]
In CB they are using Db fixtures, how is your tests implementation different from CB approach?

File: src/Quizzer.DAL.Tests/RepositoryTests.cs (hunk 2 above)
Comment (brtkkndrl) [reply src/Quizzer.DAL.Tests/RepositoryTests.cs]: It was duplicit project, removed.

File: src/Quizzer.API.DAL/Repository/Repository.cs
Comment (brtkkndrl) [reply src/Quizzer.API.DAL/Repository/Repository.cs:1]: Because we are not derriving from this class, and repositories will be only in api.

File: src/Quizzer.API.DAL.Tests/ApiDALTests.cs (hunk 1 above)
Comment (brtkkndrl) [reply src/Quizzer.API.DAL.Tests/ApiDALTests.cs]: yes

File: src/Quizzer.API.DAL/UnitOfWork/IUnitOfWork.cs (hunk 3 above)
Comment (brtkkndrl) [reply src/Quizzer.API.DAL/UnitOfWork/IUnitOfWork.cs]: So it's unnecessary in interface? What about in implementation?

**Comment by brtkkndrl** [comment]
> In CB they are using Db fixtures, how is your tests implementation different from CB approach?

I didn't see the iw5 CB implementation originally. Now looking at it in API.DAL.IntegrationTests they are using in memory repositories with `Storage`. We have `Storage` implemented, but we are using regular repositories with dbContext.

File: src/Quizzer.API.DAL/UnitOfWork/IUnitOfWork.cs (hunk 3 above)
Comment (vecerap) [reply src/Quizzer.API.DAL/UnitOfWork/IUnitOfWork.cs]: `IUnitOfWork` should inherit `IAsyncDisposable` interface which contains `DisposeAsync`, during runtime, GC also checks all instances which inherits dispose async and when the object lifetime is at it's end, it will call it implicitly...

**Comment by vecerap** [comment]
> > In CB they are using Db fixtures, how is your tests implementation different from CB approach?
> 
> I didn't see the iw5 CB implementation originally. Now looking at it in API.DAL.IntegrationTests they are using in memory repositories with `Storage`. We have `Storage` implemented, but we are using regular repositories with dbContext.

@BRTEK007  From the lectures this kind of praktice is not a good idea as you are using database during unit tests, the logic should be tested isolated from database, you now have kind of like an integration test, because you are testing also the database connection and proper working with the repository, but for the single test I would not mind it. I am not sure whether during the evaluation if they mind this wrong approach. Consider the approach from the lectures...

File: src/Quizzer.API.DAL/UnitOfWork/IUnitOfWork.cs [hunk 4]
```diff
@@ -11,10 +13,25 @@ public interface IUnitOfWork : IDisposable {
...
+
         /// <summary>
         /// Makes changes on entities in database
         /// </summary>
         void Commit();
+
+        /// <summary>
+        /// Discards changes
+        /// </summary>
+        public void Dispose();
```
Comment (vecerap) [inline src/Quizzer.API.DAL/UnitOfWork/IUnitOfWork.cs]: This method hides the Dispose() member from IDisposable, remove it.

**Comment by vecerap** [comment]
Squash commits

**Comment by vecerap** [comment]
You have conflicts, resolve them

File: src/Quizzer.API.DAL/UnitOfWork/IUnitOfWork.cs (hunk 4 above)
Comment (brtkkndrl) [reply src/Quizzer.API.DAL/UnitOfWork/IUnitOfWork.cs]: removed.

File: src/Quizzer.API.DAL/UnitOfWork/IUnitOfWork.cs (hunk 3 above)
Comment (brtkkndrl) [reply src/Quizzer.API.DAL/UnitOfWork/IUnitOfWork.cs]: removed.

**Comment by brtkkndrl** [comment]
> > > In CB they are using Db fixtures, how is your tests implementation different from CB approach?
> > 
> > 
> > I didn't see the iw5 CB implementation originally. Now looking at it in API.DAL.IntegrationTests they are using in memory repositories with `Storage`. We have `Storage` implemented, but we are using regular repositories with dbContext.
> 
> @BRTEK007 From the lectures this kind of praktice is not a good idea as you are using database during unit tests, the logic should be tested isolated from database, you now have kind of like an integration test, because you are testing also the database connection and proper working with the repository, but for the single test I would not mind it. I am not sure whether during the evaluation if they mind this wrong approach. Consider the approach from the lectures...

Testing against dbContext gives you more real scenario, I could use inMemoryDatabase but it's not guaranteed to work the same as real database. I think he mentioned on lecture that you have to assume the external code like dbContext and database are working correctly. I am not sure if this is unit or Integration so I didn't name it like that to be safe. It's kinda hacky cause we are using Storage from IW5 but with database like in ICS.

**Comment by vecerap** [comment]
> > > > In CB they are using Db fixtures, how is your tests implementation different from CB approach?
> > > 
> > > 
> > > I didn't see the iw5 CB implementation originally. Now looking at it in API.DAL.IntegrationTests they are using in memory repositories with `Storage`. We have `Storage` implemented, but we are using regular repositories with dbContext.
> > 
> > 
> > @BRTEK007 From the lectures this kind of praktice is not a good idea as you are using database during unit tests, the logic should be tested isolated from database, you now have kind of like an integration test, because you are testing also the database connection and proper working with the repository, but for the single test I would not mind it. I am not sure whether during the evaluation if they mind this wrong approach. Consider the approach from the lectures...
> 
> Testing against dbContext gives you more real scenario, I could use inMemoryDatabase but it's not guaranteed to work the same as real database. I think he mentioned on lecture that you have to assume the external code like dbContext and database are working correctly. I am not sure if this is unit or Integration so I didn't name it like that to be safe. It's kinda hacky cause we are using Storage from IW5 but with database like in ICS.

@BRTEK007 We will see after evaluation whether this is a mistake, then we will decide if we need to change it

**Review by vecerap — APPROVED** [review]
Squash the last commit

**Comment by brtkkndrl** [comment]
> Squash the last commit

done
